srml/staking/reward-curve/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+extern crate proc_macro;
+
+mod log;
+
+use log::log2;
+use proc_macro::TokenStream;
+use quote::{quote, ToTokens};
+use syn::parse::{Parse, ParseStream};
+use proc_macro2::{TokenStream as TokenStream2, Span};
+use proc_macro_crate::crate_name;
+
+/// Accepts a number of expressions to create a instance of PiecewiseLinear which represents the
+/// NPoS curve (as detailed
+/// [here](http://research.web3.foundation/en/latest/polkadot/Token%20Economics/#inflation-model))
+/// for those parameters. Parameters are:
+/// - `min_inflation`: the minimal amount to be rewarded between validators, expressed as a fraction
+///   of total issuance. Known as `I_0` in the literature.
+///   Expressed in millionth, must be between 0 and 1_000_000.
+///
+/// - `max_inflation`: the maximum amount to be rewarded between validators, expressed as a fraction
+///   of total issuance. This is attained only when `ideal_stake` is achieved.
+///   Expressed in millionth, must be between min_inflation and 1_000_000.
+///
+/// - `ideal_stake`: the fraction of total issued tokens that should be actively staked behind
+///   validators. Known as `x_ideal` in the literature.
+///   Expressed in millionth, must be between 0_100_000 and 0_900_000.
```

Comment:
what do you mean by 

> must be between 0_100_000 and 0_900_000.
 
or is it a type?

Reply:
it is just a visual way to write permillionth.
I'm parsing a u32 anyway to one could write between 0.1 and 0.9
At some point I might be able to actually make a custom parser with tokens but I feel like just taking the permillions representation is ok